weblate/trans/machine/microsoft.py

```diff
@@ -178,3 +182,172 @@ def convert_language(self, language):
+                 **kwargs):
+        """Adapter to soap_req"""
+
+        response = self.soap_req(url, http_post=False, skip_auth=True, raw=True, **kwargs)
+
+        return response
+
+    def soap_status_req(self, url, http_post=False, skip_auth=False, **kwargs):
+        """Perform SOAP request with checking response status."""
+        # Perform request
+        response = self.soap_req(url, http_post, skip_auth, **kwargs)
+
+        # Check response status
+        if response.code != 200:
+            raise MachineTranslationError(response.msg)
+
+        # Return data
+        return response
+
+    def json_status_req(self, url, http_post=False, skip_auth=False, **kwargs):
+        """Perform SOAP request with checking response status."""
+        # Perform request
+        response = self.soap_status_req(url, http_post, skip_auth=True, **kwargs)
+
+        # Check response status
+        if response.code != 200:
+            raise MachineTranslationError(response.msg)
+
+        # Return data
+        return response
+
+    def download_languages(self):
+        """List of supported languages."""
+        soap_action = 'GetLanguages'
+        soap_target = 'GetLanguagesResult'
+        soap_target_envelop = 'Code'
+        languages = []
+        xpath = self.MS_TM_XPATH
+        resp = self.soap_status_req(self.MS_TM_API_URL,
+                                    soap_action=soap_action)
+        root = ET.fromstring(resp.read())
+        results = root.find(xpath + soap_target)
+        if results is not None:
+            for lang in results:
+                languages.append(lang.find(xpath + soap_target_envelop).text)
+        return languages
+
+    def download_translations(self, source, language, text, unit, user):
+        """Download list of possible translations from a service."""
+        soap_action = 'GetTranslations'
+        soap_target = 'GetTranslationsResult'
+        soap_target_translated = 'TranslatedText'
+        soap_target_confidence = 'ConfidenceLevel'
+        soap_target_original = 'OriginalText'
+        translations = []
+        xpath = self.MS_TM_XPATH
+        resp = self.soap_status_req(self.MS_TM_API_URL,
+                                    soap_action=soap_action,
+                                    source=source,
+                                    language=language,
+                                    text=text)
+        root = ET.fromstring(resp.read())
+        results = root.find(xpath + soap_target)
+        if results is not None:
+            for translation in results:
+                translations.append(tuple([
+                    translation.find(xpath + soap_target_translated).text,
+                    int(translation.find(xpath + soap_target_confidence).text),
+                    self.name,
+                    translation.find(xpath + soap_target_original).text
+                ]))
+        return translations
+
+    def convert_language(self, language):
```

Comment:
#1673 Terminology API only supports languages like `en-us` or `en-gb` but no simple notation like `en` or `es`.  So this brings an issue when using `en` or `es` (without localization) in an effort to solve this `convert_language()` search for a matching language with localization and returns the first it founds in the list of supported languages. as an example when using `en`  `convert_language()` will return `en-au`. Any comments?

Reply:
It's probably better to use list of default locales we already maintain for the expansion:

https://github.com/WeblateOrg/weblate/blob/12dbb38dc2fbbc4bf5f52bd0c55bf385f4f1908a/weblate/lang/data.py#L931-L1036

Reply:
Awesome, will do